compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/LimitScatterBounds.cpp

```diff
@@ -0,0 +1,281 @@
+             ValueRange{empty}, ArrayRef<AffineMap>{inMap, outMap}, iters,
+             [](OpBuilder &bb, Location loc, ValueRange args) {
+               bb.create<linalg::YieldOp>(loc, args[0]);
+             })
+            .getResult(0);
+  }
+
+  llvm::SmallVector<Value> newInputs(op.getInputs());
+
+  ShapedType inputTy = cast<ShapedType>(newInputs.front().getType());
+  Value emptyBool =
+      b.create<tensor::EmptyOp>(inputTy.getShape(), b.getI1Type());
+  newInputs.push_back(emptyBool);
+
+  llvm::SmallVector<Value> newUpdates(op.getUpdates());
+  newUpdates.push_back(inbounds);
+
+  llvm::SmallVector<Type> newResultTys(op.getResultTypes());
+  newResultTys.push_back(emptyBool.getType());
+
+  auto scatter = b.create<mlir::stablehlo::ScatterOp>(
+      newResultTys, newInputs, bounded, newUpdates, dimnums,
+      op.getIndicesAreSorted(), op.getUniqueIndices());
+
+  // Replace the existing yield with conditional results and the boolean update:
+  for (int i = 0; i < op.getNumResults(); ++i) {
+    op.getResult(i).replaceAllUsesWith(scatter.getResult(i));
+  }
+
+  // Clone the region into the new scatter and insert the new boolean udpates:
+  Region &srcRegion = op.getUpdateComputation();
+  Region &dstRegion = scatter.getUpdateComputation();
+
+  IRMapping mapping;
+  srcRegion.cloneInto(&dstRegion, mapping);
+
+  const int64_t numUpdates = op.getUpdates().size();
+  dstRegion.front().insertArgument(
+      numUpdates * 2, RankedTensorType::get({}, b.getI1Type()), op.getLoc());
+  dstRegion.front().insertArgument(
+      numUpdates, RankedTensorType::get({}, b.getI1Type()), op.getLoc());
+
+  // Update the return to only conditionally return the new value:
+  Operation *terminator = dstRegion.front().getTerminator();
+  b.setInsertionPoint(terminator);
+
+  Value argInbounds = dstRegion.front().getArguments().back();
+  llvm::SmallVector<Value> terminatorOperands;
+  for (int i = 0; i < terminator->getNumOperands(); ++i) {
+    auto newArg =
+        b.create<arith::SelectOp>(argInbounds, terminator->getOperand(i),
+                                  dstRegion.front().getArgument(i));
+
+    terminatorOperands.push_back(newArg);
+  }
+
+  terminatorOperands.push_back(argInbounds);
+  b.create<mlir::stablehlo::ReturnOp>(op.getLoc(), terminatorOperands);
+
+  // Cleanup the no longer needed operations:
+  terminator->erase();
+  op.erase();
+
+  return success();
+}
+
+struct LimitScatterBounds final
+    : impl::LimitScatterBoundsBase<LimitScatterBounds> {
+
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, mlir::arith::ArithDialect,
+                    tensor::TensorDialect>();
+  }
+
+  void runOnOperation() override {
+    auto operation = getOperation();
+    llvm::SmallVector<mlir::stablehlo::ScatterOp> scatters;
+    operation.walk([&scatters](mlir::stablehlo::ScatterOp scatter) {
+      scatters.push_back(scatter);
```

Comment:
OOC why not just handle it here vs populating?  (walk does early inc range)

Reply:
We need to replace all `stablehlo.scatters` with a bounded variant. I decided to populate and replace to guarantee we do not replace the same operation twice. I also thought performing replacements during a walk can invalidate the op list in the block causing a crash?